library/alloc/src/vec.rs

```diff
@@ -2928,6 +2934,37 @@ impl<T> IntoIter<T> {
         self.ptr = self.buf.as_ptr();
         self.end = self.buf.as_ptr();
     }
+
+    /// Shifts the remaining elements to `offset` and then converts the whole allocation into a Vec
+    /// with `vec.len() == offset + self.len()`
+    ///
+    /// # Safety
+    ///
+    /// When a non-zero offset is passed the resulting Vec will have an uninitialized prefix
+    /// that needs to be filled before the Vec is valid again.
+    ///
+    /// * `offset + self.len()` must not exceed `self.cap`
+    /// * `offset == 0` is always valid
+    /// * `offset` must be positive
+    unsafe fn into_vec_with_uninit_prefix(self, offset: isize) -> Vec<T> {
+        debug_assert!(offset > 0);
+        debug_assert!(offset as usize + self.len() <= self.cap);
+        let dst = unsafe { self.buf.as_ptr().offset(offset) };
+        if self.ptr != dst as *const _ {
```

Comment:
If we disallow `0` as an input then we shouldn't need this branch.

Reply:
No, it would still be needed to shift things to the front if `IntoIter` has been advanced. All that this changes is that things can be shifted to another position than the 0 index.

Reply:
Well, we could duplicate the stuff since one version performs slightly different, it might perform better since it should be easier to optimize?

Reply:
I don't think so, the 0 passed is constant, so all the differences should optimize away and they should only amount to a handful of instructions anyway. Plus only having one method (only generic over `T`) results in slightly less llvm IR.

Reply:
Should we add a test to make sure this branch is optimized out? I don't know how this works though.

Reply:
Uh, there might be some confusion here.

In the general case passing 0 is not expected to eliminate the branch. Only in the special case where the compiler can prove that `IntoIter` has not been advanced and the target offset is 0, then we expect it not to be taken.

Reply:
I have added a comment to the branch that hopefully clears up this confusion. The debug assert above giving the wrong impression probably didn't help either.